### ISSUE TITLE
GPU memory required for training Yolo-v3

### ISSUE THREAD

**Issue by wureka** [AlexeyAB/darknet#636]
I just trained yolo-v3 for 7 classes. My hardware is JETSON TX2.
I followed the instructions to change classes and filters for  each of 3 [yolo]-layers and 3[convolutional] layers before [yolo] as below
```
[convolutional]
filters=36

[yolo]
classes=7
```
I also replaced anchors of 3 [yolo]-layers with the result of 
` ./darknet detector calc_anchors ../cfg/ai.640x384.data -num_of_clusters 9 -width 640 -heigh 384`

```
 batch= 32
 subdivisions= 32
width=640
height=384
```
The training log was below:
```
Region 94 Avg IOU: 0.190856, Class: 0.831693, Obj: 0.542294, No Obj: 0.559379, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 106 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.475702, .5R: nan, .75R: nan,  count: 0
Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.487782, .5R: nan, .75R: nan,  count: 0
Region 94 Avg IOU: 0.126236, Class: 0.129014, Obj: 0.656850, No Obj: 0.560546, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 106 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.473931, .5R: nan, .75R: nan,  count: 0
Region 82 Avg IOU: 0.155751, Class: 0.728109, Obj: 0.567061, No Obj: 0.489448, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 94 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.560938, .5R: nan, .75R: nan,  count: 0
Region 106 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.471869, .5R: nan, .75R: nan,  count: 0
Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.489748, .5R: nan, .75R: nan,  count: 0
Region 94 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.561954, .5R: nan, .75R: nan,  count: 0
Region 106 Avg IOU: 0.145725, Class: 0.555193, Obj: 0.632314, No Obj: 0.472896, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.488246, .5R: nan, .75R: nan,  count: 0
Region 94 Avg IOU: 0.392975, Class: 0.400324, Obj: 0.533230, No Obj: 0.560875, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 106 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.472750, .5R: nan, .75R: nan,  count: 0
Region 82 Avg IOU: 0.428216, Class: 0.651273, Obj: 0.224361, No Obj: 0.488955, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 94 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.562515, .5R: nan, .75R: nan,  count: 0
Region 106 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.472054, .5R: nan, .75R: nan,  count: 0
Region 82 Avg IOU: 0.093910, Class: 0.450770, Obj: 0.548225, No Obj: 0.489352, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 94 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.558836, .5R: nan, .75R: nan,  count: 0
Region 106 Avg IOU: 0.314969, Class: 0.234830, Obj: 0.538750, No Obj: 0.473070, .5R: 0.000000, .75R: 0.000000,  count: 1

 2: 3126.717285, 3125.974609 avg, 0.000000 rate, 31.244499 seconds, 64 images
```
And from JETSON TX2's log, I found that yolov3 uses GPU memory over 4GB
```
RAM 5846/7851MB (lfb 86x4MB) cpu [0%@345,0%@345,0%@345,1%@345,0%@345,45%@345]
RAM 5846/7851MB (lfb 86x4MB) cpu [21%@806,0%@345,0%@345,6%@806,0%@806,9%@806]
RAM 5846/7851MB (lfb 86x4MB) cpu [0%@345,0%@345,0%@345,0%@345,25%@345,0%@345]
RAM 5846/7851MB (lfb 86x4MB) cpu [1%@345,0%@345,0%@345,0%@345,34%@345,0%@345]
RAM 5846/7851MB (lfb 86x4MB) cpu [7%@345,0%@345,0%@345,33%@345,0%@345,6%@345]
RAM 5846/7851MB (lfb 86x4MB) cpu [0%@345,0%@345,0%@345,0%@345,0%@345,35%@345]
RAM 5846/7851MB (lfb 86x4MB) cpu [0%@345,0%@345,0%@345,0%@345,0%@345,49%@345]
```

My questions are:
1. from training log, why are some Avg IOU's value nan?
2. According to your instructions, you say yolo v3 requires 4GB GPU RAM. buy why does it take over 4GB on my environment?
Are there any configuration I could miss ?
Thanks.

**Comment by AlexeyAB** [comment]
1. `nan` occurs when there isn't labels (box truth) for anchors in this `[yolo]`-layer (`anchor=` which are specified in `mask=` in cfg-file for the current one of three yolo-layers)

2. `nan` occurs when there isn't labels for this image at all (negative samples).

Only if `nan` occurs for several dozen iterations, then training went wrong. Otherwise, the training goes well.

----

Yolo does in a loop for each Ground truth labels:
1. Yolo looking for the most suitable anchor for the current Ground truth labels (from all 9 anchors): https://github.com/AlexeyAB/darknet/blob/5c1e8e3f48343d8944af1195e21f6f3b53ed848e/src/yolo_layer.c#L243-L245
2. Then Yolo check is this the best anchors specified in `mask=` of **this `[yolo]`-layers** (in total there are 3 yolo-layers): https://github.com/AlexeyAB/darknet/blob/5c1e8e3f48343d8944af1195e21f6f3b53ed848e/src/yolo_layer.c#L249
3. And if yes, then calculates these indicators - i.e. calculates sum and divide it on number of Ground truth labels which suitable for anchor from this `[yolo]`-layer:

* `Region 82` - index of current [yolo]-layer (in the default `yolov3.cfg` there are 3 yolo layers: 82, 94, 106)
* `Avg IOU` - is average intersect of union between Detected objects and Ground truth from label-txt-file
* `Class:` - is average of the probabilities of correctly classified objects
* `Obj:` - is average of the objectness `T0` (probabilities that there is an object in this cell)
* `.5R:` - is average true positives with (IoU > 0.5)
* `.75R:` - is average true positives with (IoU > 0.75)

**Conclusion**: So if the best anchor isn't suitable for this `[yolo]`-layer, then `count` of Ground truth labels which suitable for this `[yolo]`-layer is equal to **zero** (`count=0`), then will be divide by zero
* `Avg IOU` = <avg_iou>/count = <avg_iou>/0 = `nan`
* `Class` = <avg_cat>/class_count = <avg_cat>/0 = `nan`
* `Obj` = <avg_obj>/count = <avg_obj>/0  = `nan`

```
Region 82 Avg IOU: 0.093910, Class: 0.450770, Obj: 0.548225, No Obj: 0.489352, .5R: 0.000000, .75R: 0.000000,  count: 1
Region 94 Avg IOU: nan, Class: nan, Obj: nan, No Obj: 0.558836, .5R: nan, .75R: nan,  count: 0
Region 106 Avg IOU: 0.314969, Class: 0.234830, Obj: 0.538750, No Obj: 0.473070, .5R: 0.000000, .75R: 0.000000,  count: 1
```

----

But this indicator calculated in any cases:
* `No Obj:` - is average of the probabilities of all objects (both correctly and not correctly classified)

**Comment by AlexeyAB** [comment]
> According to your instructions, you say yolo v3 requires 4GB GPU RAM. buy why does it take over 4GB on my environment?
> Are there any configuration I could miss ?

Yolo v3 requires **4 GB GPU** RAM + **~2 GB CPU** RAM ~= 6 GB RAM.

But Jetson TX2 has **only one type of memory** LPDDR4 that shared across CPU and GPU, so total spent 6 GB RAM: https://devtalk.nvidia.com/default/topic/1002349/jetson-tx2/jetson-tx2-gpu-memory-/

**Comment by wureka** [comment]
@AlexeyAB Thank you for your answer.
Regarding the portion of nan you mentioned above:

1. nan occurs when there isn't labels (box truth) for anchors in this [yolo]-layer (anchor= which are specified in mask= in cfg-file for the current one of three yolo-layers)
2. nan occurs when there isn't labels for this image at all (negative samples).

With the same dataset, if I trained with Yolo v2, there was no nan appearing. So, could I say that my dataset and relevant stuff such labels  have no problems ?

**Comment by AlexeyAB** [comment]
> So, could I say that my image dataset and relevant stuff such as labels have no problems and can be trained by Yolo v3 ?

Yes, if you can train about 6000 iterations and can get good mAP.

**Comment by anguoyang** [comment]
@wureka according to the instruction of @AlexeyAB , the width and height should be multiple of 32, is it really no problem when you training on your dataset?
My training dataset is 640x480, but I used 416x416 in the cfg file, I am not sure which one is better(use the original height width or multiple of 32)

**Comment by wureka** [comment]
@anguoyang all images of my dataset is 640x384, which are also the multiple of 32, and I also change the width and height in cfg to 640x384 classes, filters, and  anchors. So, I think they should be no problem

**Comment by TaihuLight** [comment]
@AlexeyAB When I train YOLOv3 on VOC2012+2007, and only set the different parameters as follow:
# Training , takes up 6906MiB GPU RAM
batch=128
subdivisions=32

# Training  only takes up  3876MiB GPU RAM, and lower than above setting, why?
batch=256
subdivisions=32

@AlexeyAB @wureka  @anguoyang What is wrong?

 7: 597.089417, 597.692444 avg, 0.000000 rate, 13.913467 seconds, 1792 images
Loaded: 0.000018 seconds
Region 82 Avg IOU: 0.464548, Class: 0.635690, Obj: 0.374416, No Obj: 0.521148, .5R: 0.500000, .75R: 0.000000,  count: 6
Region 94 Avg IOU: 0.392822, Class: 0.381941, Obj: 0.574848, No Obj: 0.554483, .5R: 0.272727, .75R: 0.000000,  count: 11
Region 106 Avg IOU: 0.367993, Class: 0.437119, Obj: 0.463624, No Obj: 0.506273, .5R: 0.250000, .75R: 0.000000,  count: 4
Region 82 Avg IOU: 0.325769, Class: 0.561967, Obj: 0.457437, No Obj: 0.519700, .5R: 0.250000, .75R: 0.000000,  count: 4
Region 94 Avg IOU: 0.217476, Class: 0.340199, Obj: 0.570449, No Obj: 0.554392, .5R: 0.000000, .75R: 0.000000,  count: 2
Region 106 Avg IOU: -nan, Class: -nan, Obj: -nan, No Obj: 0.506313, .5R: -nan, .75R: -nan,  count: 0
Region 82 Avg IOU: 0.337817, Class: 0.508326, Obj: 0.555578, No Obj: 0.520447, .5R: 0.166667, .75R: 0.000000,  count: 6
Region 94 Avg IOU: 0.379297, Class: 0.529548, Obj: 0.584318, No Obj: 0.554746, .5R: 0.000000, .75R: 0.000000,  count: 3
Region 106 Avg IOU: -nan, Class: -nan, Obj: -nan, No Obj: 0.506114, .5R: -nan, .75R: -nan,  count: 0
'
'

**Comment by AlexeyAB** [comment]
@TaihuLight 

Do you train with `random=1`?
Did you see this message  `CUDNN-slow` during training?

If there is not enough GPU-memory, then will be used slower cuDNN algorithm that doesn't require extra spaces on GPU.

**Comment by viethoang303** [comment]
@AlexeyAB thank you for this analysis, so how to fix this??? it's simply training with multiple epochs, right????. Maybe I'm wrong
